### ISSUE TITLE
React prop validation not supported

### ISSUE THREAD

**Issue by pelotom** [amilajack/eslint-plugin-flowtype-errors#14]
This code:

```javascript
// @flow

import React, { Component } from 'react'

class Foo extends Component {
    props: {
        firstName: string,
        lastName: string,
    }

    render() {
        return (
            <div>
                <h1>Hello {this.props.firstName} {this.props.lastName}</h1>
            </div>
        )
    }
}

class Bar extends Component {
    render() {
        return <Foo firstName="John" />
    }
}
```

rightly fails typechecking with `flow` because `Bar` is not supplying all required props to `Foo`:

```
  6:     props: {
                ^ property `lastName`. Property not found in
 22:         return <Foo firstName="John" />
                    ^^^^^^^^^^^^^^^^^^^^^^^^ props of React element `Foo`
```

however `eslint-plugin-flowtype-errors` doesn't notice the problem.

**Comment by amilajack** [comment]
Thanks for opening this issue! I'll start looking into this right now.

**Comment by amilajack** [comment]
<img width="996" alt="screen shot 2016-09-30 at 9 46 40 pm" src="https://cloud.githubusercontent.com/assets/6374832/19011824/6393b32a-8757-11e6-896b-8fbf2f4f6909.png">

Is this the expected output?

**Comment by pelotom** [comment]
Ideally the error should appear in `Bar`, at the use site where `Foo` is being given the wrong props, not in `Foo` itself.

**Comment by amilajack** [comment]
<img width="627" alt="screen shot 2016-09-30 at 10 13 52 pm" src="https://cloud.githubusercontent.com/assets/6374832/19011958/30c28774-875b-11e6-856c-5ac5cd83a545.png">

Unfortunately that's just how Flow formats it :(

**Comment by amilajack** [comment]
Their wording is really tricky. Do you suggest I duplicate the message at both lines 6 and 22?

So maybe something like this?
```
6: property `lastName`. Property not found in props of React element `Foo`
21: property `lastName`. Property not found in props of React element `Foo`
```

**Comment by pelotom** [comment]
Yeah, I figured, because every IDE package for flow seems to put the error there. Looks like a bug I need to take up with the Flow team! This is great though, thanks for your prompt fix!

**Comment by amilajack** [comment]
I think this is intentional. They're basically trying to say what the error is and then try to point to it (which is useful but I think they could improve on this). The current way they achieve this makes it impossible for IDE's/plugins to highlight both with points with their own error messages.

**Comment by amilajack** [comment]
Before I release the fixesI want to check if this address all the issues you saw:

You can install the testing version like so and it should work
```
npm i -D eslint-plugin-flowtype-errors@1.2.4-0
```

**Comment by pelotom** [comment]
Yep, it works! 👍

**Comment by pelotom** [comment]
> The current way they achieve this makes it impossible for IDE's/plugins to highlight both with points with their own error messages.

I realize it's not possible to highlight both, but is it possible to highlight only the more relevant one? The information needed seems to be there in the console output from `flow`:

```
src/components/foo.js:6
  6:     props: {
                ^ property `lastName`. Property not found in
 22:         return <Foo firstName="John" />
                    ^^^^^^^^^^^^^^^^^^^^^^^^ props of React element `Foo`
```

The first part of that error message is what's being highlighting right now, but the second part, the use site, is the interesting thing and what should ideally be used for the error marker location. I'm not familiar with how Flow is used as an API, but if you have access to the second location, that would be much more useful.

**Comment by pelotom** [comment]
It looks like from the JSON API on [Try Flow](https://flowtype.org/try/#0PTAEAEDMBsHsHcBQiCWBbADrATgF1AEoCmAhgMa4A0oA3qAMKyawB2RL+AvqJNk6AHJspCgORloJAM5TQAMVixQRAB652AE1mNmbDrUSgjoDHwxSAXAeM2eKbFNwA5EmiJXH2FCwDmlQ7ZGko4ubh64Xr7+NpzINsIsGkTYABQAlNaBoMK4AK7YLKApAVnGADwaKABuAHwlpbZlABYAjDUAEkTQcLS4TShSAHSmsOaDkPYhrkTcNH0Dw2ZDwc7TnGXArXUN5cCVtfU2afWxsYgS0rIAQiTYymqa2kxYevg09QlJqRnvpTn5hTKCiUEwcqzcAF4AEQAKVgTRYUNAwG2xlOQA) that the relevant information is there in the third context block:

```json
{
  "context": "        return <Foo firstName=\"John\" />",
  "descr": "props of React element `Foo`",
  "type": "Blame",
  "loc": {
    "source": "-",
    "type": "SourceFile",
    "start": {
      "line": 22,
      "column": 16,
      "offset": 379
    },
    "end": {
      "line": 22,
      "column": 39,
      "offset": 403
    }
  },
  "path": "-",
  "line": 22,
  "endline": 22,
  "start": 16,
  "end": 39
}
```

**Comment by pelotom** [comment]
This definitely seems like an issue to bring up with Flow though, because they're reversing the normal ordering of relevancy when it comes to prop errors vs function calls. With errors in function arguments, the first context is the relevant one (the call site), whereas with React element creation the first context is the declaration site, and the "call site" comes last. I will bring this up with the Flow project.

**Comment by amilajack** [comment]
> This definitely seems like an issue to bring up with Flow though

Yup. That's what I was thinking. Should this be closed then?

**Comment by pelotom** [comment]
Yep, thanks!

**Comment by amilajack** [comment]
No problem! Also can I get your answer to [this question](https://github.com/amilajack/eslint-plugin-flowtype-errors/issues/8#issuecomment-250891205). Some editors dont seem to support linting before saving the file.

**Comment by pelotom** [comment]
I filed [an issue for this](https://github.com/facebook/flow/issues/2557) with the Flow project.

**Comment by amilajack** [comment]
Great! Thanks a bunch.
